### Домашнее задание 4 - 10 баллов

В этом задании вам предстоит дообучить трансформерную модель для NER-задачи в различных форматах:

1. Обучите NER-модель

- Загрузите набор данных [Collection5](https://github.com/natasha/corus?tab=readme-ov-file#load_ne5) - **1 балл**
- Разбейте набор данных на train/test части
- Дообучите модель [rubert-tiny2](https://huggingface.co/cointegrated/rubert-tiny2) на train-части корпуса для решения NER-задачи, сделайте замеры качества NER-метрик до и после дообучения - **2 балла**

2. Попробуйте улучшить качество модели следующими способами:
- Предварительно дообучите на train-части в MLM режиме, а потом дообучите на NER-задачу - **2 балла**
- Сгенерируйте синтетическую разметку* подходящего**, на ваш взгляд, новостного корпуса большой и умной моделью для русскоязычного NER***, а затем использовав ее для дообучения rubert-tiny2 вместе с основным набором данных - **2 балла**

3. Финально сравните результаты различных подходов - **1 балл**

*прогоните датасет через NER-модель, получите ее предсказания и используйте их в качестве резметки

**Можно использовать уже знакомый вам датасет lenta-ru, объем данных лучше взять от 10_000 текстов

***Например, можно взять модель модель DeepPavlov ner_collection3_bert. Инструкция по запуску есть в [документации](https://docs.deeppavlov.ai/en/master/features/models/NER.html)

In [ ]:
!pip install corus

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 4.3 MB/s eta 0:00:00


In [ ]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.2/491.2 kB 12.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 9.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.9/183.9 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 17.0 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.2
    Uninstalling fsspec-2025.3.2:
      Successfully uninstalled fsspec-2025.3.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.2 requires fsspec==2025.3.2, but you have fsspec 2024.12.0 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which 

In [ ]:
!pip install evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 3.5 MB/s eta 0:00:00


In [ ]:
from corus import load_ne5, load_lenta
import corus.sources.ne5 as ne5
import torch
import numpy as np
import pandas as pd
from tqdm import tqdm
import random

from datasets import Dataset, DatasetDict, concatenate_datasets
from transformers import AutoTokenizer, AutoModelForTokenClassification, Trainer, TrainingArguments, DataCollatorForTokenClassification
from transformers import AutoModelForMaskedLM, DataCollatorForLanguageModeling, EarlyStoppingCallback, pipeline
from sklearn.model_selection import train_test_split

from collections import defaultdict
import evaluate
import tabulate

In [ ]:
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

**Загрузка и подготовка набора данных Collection5**

In [ ]:
!wget http://www.labinform.ru/pub/named_entities/collection5.zip

--2025-04-16 06:58:27--  http://www.labinform.ru/pub/named_entities/collection5.zip
Resolving www.labinform.ru (www.labinform.ru)... 95.181.230.181
Connecting to www.labinform.ru (www.labinform.ru)|95.181.230.181|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1899530 (1.8M) [application/zip]
Saving to: ‘collection5.zip’

collection5.zip     100%[===================>]   1.81M  1.50MB/s    in 1.2s    

2025-04-16 06:58:29 (1.50 MB/s) - ‘collection5.zip’ saved [1899530/1899530]



In [ ]:
!unzip collection5.zip

Archive:  collection5.zip
   creating: Collection5/
  inflating: Collection5/001.ann     
  inflating: Collection5/001.txt     
  inflating: Collection5/002.ann     
  inflating: Collection5/002.txt     
  inflating: Collection5/003.ann     
  inflating: Collection5/003.txt     
  inflating: Collection5/004.ann     
  inflating: Collection5/004.txt     
  inflating: Collection5/005.ann     
  inflating: Collection5/005.txt     
  inflating: Collection5/006.ann     
  inflating: Collection5/006.txt     
  inflating: Collection5/007.ann     
  inflating: Collection5/007.txt     
  inflating: Collection5/008.ann     
  inflating: Collection5/008.txt     
  inflating: Collection5/009.ann     
  inflating: Collection5/009.txt     
  inflating: Collection5/010.ann     
  inflating: Collection5/010.txt     
  inflating: Collection5/011.ann     
  inflating: Collection5/011.txt     
  inflating: Collection5/012.ann     
  inflating: Collection5/012.txt     
  inflating: Collection5/013.ann    

In [ ]:
def load_text_utf8(path):
    with open(path, 'r', encoding='utf-8') as f:
        return f.read()

ne5.load_text = load_text_utf8
data = load_ne5("Collection5")
docs = list(data)

# Собираем все типы сущностей
types = set()
for doc in docs:
    for span in doc.spans:
        types.add(span.type)

label_list = ['O']
for entity_type in sorted(types):
    label_list.extend([f'B-{entity_type}', f'I-{entity_type}'])
label2id = {label: idx for idx, label in enumerate(label_list)}

tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

def process_doc(doc):
    text = doc.text
    spans = sorted(doc.spans, key=lambda x: x.stop - x.start, reverse=True)  # Приоритет длинным

    # Токенизация
    tokenized = tokenizer(
        text,
        return_offsets_mapping=True,
        truncation=False,
        add_special_tokens=False
    )

    tokens = []
    for token, offset in zip(tokenized.tokens(), tokenized.offset_mapping):
        if token in tokenizer.all_special_tokens:
            continue
        tokens.append({
            'token': token,
            'start': offset[0],
            'end': offset[1],
            'is_word_start': not token.startswith('##')
        })

    # Инициализация меток
    token_labels = ['O'] * len(tokens)

    # Обработка спанов в порядке приоритета
    for span in spans:
        span_type = span.type
        span_start = span.start
        span_end = span.stop

        for i, token in enumerate(tokens):
            if token_labels[i] != 'O':
                continue

            if token['start'] >= span_end or token['end'] <= span_start:
                continue

            # Определение B/I
            prev_token = tokens[i-1] if i > 0 else None
            if prev_token and (prev_token['end'] > span_start):
                prefix = 'I-'
            else:
                prefix = 'B-'

            if token['is_word_start']:
                token_labels[i] = f'{prefix}{span_type}'
            else:
                token_labels[i] = f'I-{span_type}'

    return {
        'tokens': [t['token'] for t in tokens],
        'ner_tags': [label2id[label] for label in token_labels]
    }

# Обрабатываем все документы
processed_data = [process_doc(doc) for doc in docs]

# Создаем Dataset
dataset = Dataset.from_list(processed_data)
dataset.features['ner_tags'].feature.names = label_list

print(dataset)
print(dataset[0])
print(f"{label_list=}")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/401 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.74M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Token indices sequence length is longer than the specified maximum sequence length for this model (2392 > 2048). Running this sequence through the model will result in indexing errors


Dataset({
    features: ['tokens', 'ner_tags'],
    num_rows: 1000
})
{'tokens': ['Д', '.', 'Медведев', 'снял', 'с', 'должности', 'зам', '##секрет', '##аря', 'Совбеза', 'РФ', 'Ю', '.', 'Бал', '##у', '##евского', 'Президент', 'России', 'Дмитрий', 'Медведев', 'освободил', 'Юрия', 'Бал', '##у', '##евского', 'от', 'должности', 'зам', '##секрет', '##аря', 'Совета', 'безопасности', 'России', '.', 'Соответствующий', 'указ', 'опубликован', 'на', 'сайте', 'государственной', 'системы', 'правовой', 'информации', '.', 'Указ', 'глава', 'государства', 'подписал', '9', 'января', '.', 'Именно', 'в', 'этот', 'день', 'Ю', '.', 'Бал', '##у', '##евском', '##у', 'исполнилось', '65', 'лет', '.', 'Согласно', 'российскому', 'законодательству', ',', 'предел', '##ьный', 'возраст', 'для', 'нахождения', 'на', 'государственной', 'службе', 'составляет', '60', 'лет', ',', 'однако', 'он', 'может', 'быть', 'увеличен', 'до', '65', 'лет', '.', 'Дальнейшее', 'продление', 'нахождения', 'на', 'госслу', '##жбе', 'не', 'допу

**Подсчет статистики**

In [ ]:
label_counts = defaultdict(int)
entity_counts = defaultdict(int)

for example in dataset:
    for tag_id in example['ner_tags']:
        label = label_list[tag_id]
        label_counts[label] += 1
        if label.startswith('B-'):
            entity_type = label[2:]
            entity_counts[entity_type] += 1

print("\nРаспределение меток:")
for label in label_list:
    print(f"{label}: {label_counts[label]}")

print("\nКоличество сущностей по типам:")
for entity, count in sorted(entity_counts.items()):
    print(f"{entity}: {count}")


Распределение меток:
O: 229790
B-GEOPOLIT: 3414
I-GEOPOLIT: 3238
B-LOC: 2641
I-LOC: 6566
B-MEDIA: 1391
I-MEDIA: 2922
B-ORG: 5952
I-ORG: 16921
B-PER: 8185
I-PER: 30171

Количество сущностей по типам:
GEOPOLIT: 3414
LOC: 2641
MEDIA: 1391
ORG: 5952
PER: 8185


**Напишем функцию для вычисления метрик на основе предсказаний модели и истинных меток**

In [ ]:
!pip install seqeval

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 3.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16162 sha256=a56be832f150315685d9c8fcd1d5da999e2ac7e337e1f818633b9fab1d505ef1
  Stored in directory: /root/.cache/pip/wheels/bc/92/f0/243288f899c2eacdfa8c5f9aede4c71a9bad0ee26a01dc5ead
Successfully built seqeval


In [ ]:
seqeval = evaluate.load("seqeval")

def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    true_predictions = [
        [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[l] for l in label if l != -100]
        for label in labels
    ]

    results = seqeval.compute(
        predictions=true_predictions,
        references=true_labels,
        zero_division=0
    )
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"]
    }

**Дообучение модели rubert-tiny2 на train-части и оценка качества NER-метрик до и после дообучения**

In [ ]:
model_name = "cointegrated/rubert-tiny2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForTokenClassification.from_pretrained(
    model_name,
    num_labels=len(label_list),
    id2label={i: label for i, label in enumerate(label_list)},
    label2id=label2id
)

def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True,
        is_split_into_words=True,
        padding="max_length",
        max_length=128
    )

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

splitted_dataset = dataset.train_test_split(test_size=0.2, shuffle=True, seed=42) if not isinstance(dataset, DatasetDict) else dataset
tokenized_dataset = splitted_dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=splitted_dataset["train"].column_names
)


# Параметры обучения
training_args = TrainingArguments(
    output_dir="./ner_results",
    eval_strategy="steps",
    save_strategy="steps",
    save_steps=1000,
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    max_steps=10000,
    logging_steps=100,
    weight_decay=0.01,
    metric_for_best_model="f1",
    logging_dir="./logs",
    report_to="none",
    load_best_model_at_end=True,
    save_total_limit=2,
    )

# Дообучение
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer),
    compute_metrics=compute_metrics,
    )

trainer.add_callback(EarlyStoppingCallback(
    early_stopping_patience=10,
    early_stopping_threshold=0.001,
    ))

baseline_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ДО дообучения:")
print(tabulate.tabulate(
    baseline_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))


trainer.train()

final_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ПОСЛЕ дообучения:")
print(tabulate.tabulate(
    final_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

config.json:   0%|          | 0.00/693 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/118M [00:00<?, ?B/s]

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Метрики ДО дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.4717 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0015 |
+-----------------------------+------------+
| eval_precision              |     0.0019 |
+-----------------------------+------------+
| eval_recall                 |     0.0162 |
+-----------------------------+------------+
| eval_f1                     |     0.0034 |
+-----------------------------+------------+
| eval_accuracy               |     0.0371 |
+-----------------------------+------------+
| eval_runtime                |     1.5858 |
+-----------------------------+------------+
| eval_samples_per_second     |   126.1220 |
+-----------------------------+------------+
| eval_steps_per_second       |     8.1980 |
+-----------------------------+------------+


Метрики ПОСЛЕ дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.3710 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0015 |
+-----------------------------+------------+
| eval_precision              |     0.5028 |
+-----------------------------+------------+
| eval_recall                 |     0.5545 |
+-----------------------------+------------+
| eval_f1                     |     0.5274 |
+-----------------------------+------------+
| eval_accuracy               |     0.9129 |
+-----------------------------+------------+
| eval_runtime                |     0.6392 |
+-----------------------------+------------+
| eval_samples_per_second     |   312.8710 |
+-----------------------------+------------+
| eval_steps_per_second       |    20.3370 |
+-----------------------------+------------+
| epoch                      

**Будем оценивать F1-меру, потому что она учитывает как точность, так и полноту. Это наиболее честная оценка для несбалансированных данных:**
- F1 до дообучения: 0.0034     
- F1 после дообучения: 0.5274

Нам неплохо удалось увеличить качество!

**Попробуем улучшить качество модели, предварительное дообучив её на train-части в MLM режиме, а потом дообучив на NER-задачу**

In [ ]:
train_docs, test_docs = train_test_split(docs, test_size=0.2, shuffle=True, random_state=42)

tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

# Подготовка данных для MLM
mlm_texts = [doc.text for doc in train_docs]
mlm_dataset = Dataset.from_dict({"text": mlm_texts})

def tokenize_mlm(examples):
    return tokenizer(
        examples["text"],
        truncation=True,
        padding="max_length",
        max_length=128,
        return_special_tokens_mask=True
    )

tokenized_mlm = mlm_dataset.map(tokenize_mlm, batched=True, remove_columns=["text"])

# Обучение MLM
mlm_model = AutoModelForMaskedLM.from_pretrained("cointegrated/rubert-tiny2")
data_collator_mlm = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,
    mlm_probability=0.15
)

training_args_mlm = TrainingArguments(
    output_dir="./mlm_results",
    overwrite_output_dir=True,
    max_steps=3000,
    per_device_train_batch_size=16,
    save_steps=1000,
    save_total_limit=2,
    logging_dir='./mlm_logs',
    report_to="none"
)

trainer_mlm = Trainer(
    model=mlm_model,
    args=training_args_mlm,
    train_dataset=tokenized_mlm,
    data_collator=data_collator_mlm
)

trainer_mlm.train()
mlm_model.save_pretrained("./mlm_model")

def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True,
        is_split_into_words=True,
        padding="max_length",
        max_length=128
    )

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

# Создаем Dataset
dataset = Dataset.from_list(processed_data)
dataset.features['ner_tags'].feature.names = label_list
splitted_dataset = dataset.train_test_split(test_size=0.2, shuffle=True, seed=42)
tokenized_dataset = splitted_dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=splitted_dataset["train"].column_names
)

# Инициализация модели для NER
model = AutoModelForTokenClassification.from_pretrained(
    "./mlm_model",
    num_labels=len(label_list),
    id2label={i: label for i, label in enumerate(label_list)},
    label2id=label2id
)

training_args = TrainingArguments(
    output_dir="./ner_results",
    eval_strategy="steps",
    save_strategy="steps",
    save_steps=1000,
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    max_steps =10000,
    logging_steps=100,
    weight_decay=0.01,
    metric_for_best_model="f1",
    logging_dir="./logs",
    report_to="none",
    save_total_limit=2,
    load_best_model_at_end=True,
)


trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer),
    compute_metrics=compute_metrics
)

trainer.add_callback(EarlyStoppingCallback(
    early_stopping_patience=10,
    early_stopping_threshold=0.001
))

baseline_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ДО дообучения:")
print(tabulate.tabulate(
    baseline_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

trainer.train()

final_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ПОСЛЕ дообучения:")
print(tabulate.tabulate(
    final_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Some weights of BertForTokenClassification were not initialized from the model checkpoint at ./mlm_model and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Метрики ДО дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.3439 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0008 |
+-----------------------------+------------+
| eval_precision              |     0.0035 |
+-----------------------------+------------+
| eval_recall                 |     0.0228 |
+-----------------------------+------------+
| eval_f1                     |     0.0061 |
+-----------------------------+------------+
| eval_accuracy               |     0.1536 |
+-----------------------------+------------+
| eval_runtime                |     0.4583 |
+-----------------------------+------------+
| eval_samples_per_second     |   436.4300 |
+-----------------------------+------------+
| eval_steps_per_second       |    28.3680 |
+-----------------------------+------------+


Метрики ПОСЛЕ дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.2963 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0008 |
+-----------------------------+------------+
| eval_precision              |     0.4762 |
+-----------------------------+------------+
| eval_recall                 |     0.5170 |
+-----------------------------+------------+
| eval_f1                     |     0.4957 |
+-----------------------------+------------+
| eval_accuracy               |     0.9058 |
+-----------------------------+------------+
| eval_runtime                |     0.3505 |
+-----------------------------+------------+
| eval_samples_per_second     |   570.5690 |
+-----------------------------+------------+
| eval_steps_per_second       |    37.0870 |
+-----------------------------+------------+
| epoch                      

**Также будем оценивать на F1-меру:**
- F1 до обучения: 0.0061      
- F1 после обучения: 0.5274 -> 0.4957

Вообще я ожидал, что метрика увеличится, так как MLM по идее выступает здесь как регуляризатор, улучшающий обобщающую способность модели, но возможно я что-то не так сделал. Несколько раз прогонял, и результат немного падал.

**Генерация синтетической разметки новостного `lenta-ru-news.csv.gz` корпуса `Babelscape/wikineural-multilingual-ner` NER моделью и использование ее для дообучения rubert-tiny2 вместе с основным набором данных**

In [ ]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-04-16 07:44:56--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 140.82.116.3
Connecting to github.com (github.com)|140.82.116.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250416%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250416T074456Z&X-Amz-Expires=300&X-Amz-Signature=e9aff57b3883afa0737383b336b898d41bf4f73086d68a80f6dca3f7add86956&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-04-16 07:44:56--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-

In [ ]:
path = 'lenta-ru-news.csv.gz'
records = load_lenta(path)

data = []
for record in records:
    if record.topic is None:
        continue
    data.append({
        'title': record.title,
        'text': record.text,
        'topic': record.topic
    })
    if len(data) >= 15_000:
        break
df = pd.DataFrame(data)

**Разметка текста умной моделью и сохранение в pandas датафрейм**

In [ ]:
ner_model = pipeline(
    "ner",
    model="Babelscape/wikineural-multilingual-ner",
    aggregation_strategy="simple",
    device=0 if torch.cuda.is_available() else -1
)

def process_text(text):
    """Обрабатывает текст и извлекает N-сущности с индексами"""
    entities = ner_model(text)
    return [
        {
            "ner_token": ent["word"],
            "label": ent["entity_group"],
            "start": ent["start"],
            "end": ent["end"]
        }
        for ent in entities
    ]

tqdm.pandas()
df["ner_entities"] = df["text"].progress_apply(process_text)

result_df = pd.DataFrame({
    "text": df["text"],
    "ner_entities": df["ner_entities"]
})

result_df.to_csv("ner_dataset.csv", index=False)

config.json:   0%|          | 0.00/1.19k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/709M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/333 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0
100%|██████████| 15000/15000 [08:49<00:00, 28.34it/s]


Формирование размеченного датасета для обучения

In [ ]:
tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

def result_df_to_dataset(result_df, label2id):
    processed_data = []

    for _, row in tqdm(result_df.iterrows(), total=len(result_df)):
        text = row['text']
        entities = row['ner_entities']

        tokenized = tokenizer(
            text,
            return_offsets_mapping=True,
            truncation=False,
            add_special_tokens=False
        )

        # Собираем информацию о токенах
        tokens_info = []
        for token, offset in zip(tokenized.tokens(), tokenized.offset_mapping):
            if token in tokenizer.all_special_tokens:
                continue
            tokens_info.append({
                'token': token,
                'start': offset[0],
                'end': offset[1],
                'is_word_start': not token.startswith('##')
            })

        # Создаем список меток
        token_labels = ['O'] * len(tokens_info)

        # Сортируем сущности по длине (длинные сначала)
        sorted_entities = sorted(
            entities,
            key=lambda x: x['end'] - x['start'],
            reverse=True
        )

        # Размечаем сущности
        for ent in sorted_entities:
            ent_start = ent['start']
            ent_end = ent['end']
            ent_type = ent['label']

            # Ищем пересекающиеся токены
            for i, token in enumerate(tokens_info):
                if token_labels[i] != 'O':
                    continue

                # Проверяем пересечение токена с сущностью
                if (token['start'] >= ent_end) or (token['end'] <= ent_start):
                    continue

                # Определяем B/I
                if token['is_word_start']:
                    # Проверяем начало сущности
                    if token['start'] == ent_start:
                        prefix = 'B-'
                    else:
                        prefix = 'I-'
                else:
                    prefix = 'I-'

                # Проверяем существование метки
                full_label = f"{prefix}{ent_type}"
                if full_label not in label2id:
                    continue  # Пропускаем неизвестные метки

                token_labels[i] = full_label

        # Конвертируем метки в ID
        ner_tags = [label2id.get(label, label2id['O']) for label in token_labels]

        processed_data.append({
            'tokens': [t['token'] for t in tokens_info],
            'ner_tags': ner_tags
        })

    features = dataset.features
    new_dataset = Dataset.from_list(processed_data, features=features)

    return new_dataset

label2id = {label: idx for idx, label in enumerate(label_list)}
new_dataset = result_df_to_dataset(result_df, label2id)

print(new_dataset)

100%|██████████| 15000/15000 [01:02<00:00, 240.74it/s]


Dataset({
    features: ['tokens', 'ner_tags'],
    num_rows: 15000
})


**Подсчет статистики**

In [ ]:
label_counts = defaultdict(int)
entity_counts = defaultdict(int)

for example in new_dataset:
    for tag_id in example['ner_tags']:
        label = label_list[tag_id]
        label_counts[label] += 1
        if label.startswith('B-'):
            entity_type = label[2:]
            entity_counts[entity_type] += 1

print("\nРаспределение меток:")
for label in label_list:
    print(f"{label}: {label_counts[label]}")

print("\nКоличество сущностей по типам:")
for entity, count in sorted(entity_counts.items()):
    print(f"{entity}: {count}")


Распределение меток:
O: 3333078
B-GEOPOLIT: 0
I-GEOPOLIT: 0
B-LOC: 69014
I-LOC: 50060
B-MEDIA: 0
I-MEDIA: 0
B-ORG: 50832
I-ORG: 67733
B-PER: 63963
I-PER: 116258

Количество сущностей по типам:
LOC: 69014
ORG: 50832
PER: 63963


**Объединение с оригинальным датасетом**

In [ ]:
splitted_orig_dataset = dataset.train_test_split(test_size=0.2, shuffle=True, seed=42) if not isinstance(dataset, DatasetDict) else dataset
combined_train = concatenate_datasets([
    splitted_orig_dataset['train'],
    new_dataset
])

combined_dataset = DatasetDict({
    'train': combined_train,
    'test': splitted_orig_dataset['test']
})

print(combined_dataset)

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 15800
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 200
    })
})


**Дообучение rubert-tiny2 на новом наборе данных вместе с основным**

In [ ]:
model_name = "cointegrated/rubert-tiny2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForTokenClassification.from_pretrained(
    model_name,
    num_labels=len(label_list),
    id2label={i: label for i, label in enumerate(label_list)},
    label2id=label2id
)

def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True,
        is_split_into_words=True,
        padding="max_length",
        max_length=128
    )

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

tokenized_dataset = combined_dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=combined_dataset["train"].column_names
)

training_args = TrainingArguments(
    output_dir="./ner_upd_dataset_results",
    eval_strategy="steps",
    save_strategy="steps",
    save_steps=3000,
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    max_steps=30000,
    logging_steps=1000,
    weight_decay=0.01,
    metric_for_best_model="f1",
    logging_dir="./logs",
    report_to="none",
    load_best_model_at_end=True,
    save_total_limit=2,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer),
    compute_metrics=compute_metrics,
)

baseline_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ДО дообучения:")
print(tabulate.tabulate(
    baseline_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))


trainer.train()

final_metrics = trainer.evaluate(tokenized_dataset["test"])
print("Метрики ПОСЛЕ дообучения:")
print(tabulate.tabulate(
    final_metrics.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))


Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/15800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

Метрики ДО дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.4779 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0011 |
+-----------------------------+------------+
| eval_precision              |     0.0024 |
+-----------------------------+------------+
| eval_recall                 |     0.0183 |
+-----------------------------+------------+
| eval_f1                     |     0.0043 |
+-----------------------------+------------+
| eval_accuracy               |     0.0654 |
+-----------------------------+------------+
| eval_runtime                |     0.4547 |
+-----------------------------+------------+
| eval_samples_per_second     |   439.8670 |
+-----------------------------+------------+
| eval_steps_per_second       |    28.5910 |
+-----------------------------+------------+


Метрики ПОСЛЕ дообучения:
+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.3593 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0011 |
+-----------------------------+------------+
| eval_precision              |     0.4807 |
+-----------------------------+------------+
| eval_recall                 |     0.4987 |
+-----------------------------+------------+
| eval_f1                     |     0.4895 |
+-----------------------------+------------+
| eval_accuracy               |     0.9030 |
+-----------------------------+------------+
| eval_runtime                |     0.4085 |
+-----------------------------+------------+
| eval_samples_per_second     |   489.5400 |
+-----------------------------+------------+
| eval_steps_per_second       |    31.8200 |
+-----------------------------+------------+
| epoch                      

**Также будем оценивать на F1-меру:**   
- F1 до обучения: 0.0043      
- F1 после обучения: 0.5274 -> 0.4895

**Выводы**
- Обычное NER-дообученение дало F1-score = 0.5211
- MLM вместе с NER дали F1-score = 0.4957      
- NER-дообученение на расширенной выборке, размеченной более крупной моделью, дало F1-score = 0.4895

      
1. В подходе "MLM + NER" не удалось улучшить качество модели, оно даже немного упало. Тут конечно вопросы к моей реализации, так как ожидалось обратное.
2. Автоматическая разметка данных крупной моделью могла внести ошибки (ложные срабатывания, пропуски сущностей). Возможно, крупная модель для разметки была недостаточно точной в целевой предметной области.

Основная проблема, вероятно, связана с ограничениями модели-учителя: будучи обученной размечать только сущности LOC, ORG и PER, она не учитывала типы GEOPOLIT и MEDIA. В результате автоматически размеченная выборка лишилась этих категорий, создав сильный дисбаланс в данных. При обучении на такой выборке rubert-tiny2, не получая примеров GEOPOLIT и MEDIA, модель переобучилась на доминирующие метки (LOC, ORG, PER), что привело к деградации качества на тестовых данных — модель хуже понимала непредставленные сущности, что и объясняет самый низкий результат по сравнению с другими подходами.